lib/gruf/error.rb

```diff
@@ -130,16 +138,30 @@ def serialize
     end
 
     ##
-    # Append any appropriate errors to the gRPC call and properly update the output metadata
+    # Update the trailing metadata on the given gRPC call, including the error payload if configured
+    # to do so.
     #
     # @param [GRPC::ActiveCall] active_call The marshalled gRPC call
-    # @return [Error] Return the error itself with the GRPC::ActiveCall attached and error metadata appended
+    # @return [Error] Return the error itself after updating metadata on the given gRPC call.
+    #                 In the case of a metadata overflow error, we replace the current error with
+    #                 a new one that won't cause a low-level http2 error.
     #
     def attach_to_call(active_call)
       metadata[Gruf.error_metadata_key.to_sym] = serialize if Gruf.append_server_errors_to_trailing_metadata
-      if !metadata.empty? && active_call && active_call.respond_to?(:output_metadata)
-        active_call.output_metadata.update(metadata)
+      return self unless !metadata.empty? && active_call && active_call.respond_to?(:output_metadata)
+
+      # Check if we've overflown the maximum size of output metadata. If so,
+      # log a warning and replace the metadata with something smaller to avoid
+      # resource exhausted errors.
+      if metadata.inspect.size > MAX_METADATA_SIZE
+        code = METADATA_SIZE_EXCEEDED_CODE
+        msg = METADATA_SIZE_EXCEEDED_MSG
+        logger.warn "#{code}: #{msg} Original error: #{to_h.inspect}"
```

Comment:
Is there a chance we'd be leaking sensitive data here to logs?

Reply:
No more so than what would be sent back to the client in the error payload which is most likely logged anyway. I think the risk of additional exposure is minimal.

Reply:
Okay, the difference there though is the client (and any interceptors) have control over what is logged. This way we're making it canon and unavoidable.